From d59b6819b15ed72fd374c0ab308b1731316f93b0 Mon Sep 17 00:00:00 2001
From: Adrien Gonzalez <agonzalez@prophesee.ai>
Date: Fri, 19 Sep 2025 14:49:26 +0200
Subject: [PATCH] Updates V4L2 Biases facility

 - Defers bias range check to the V4L2 driver control by forcing bypass check range on I_LL_Bias class
 - Updates V4L2 Bias Info to properly advertise its allowed range based on sensor idac bitfield width

---
 .../devices/v4l2/v4l2_ll_biases.h             |  6 ++++-
 .../v4l2/v4l2_hardware_identification.cpp     |  2 +-
 .../src/devices/v4l2/v4l2_device_builder.cpp  | 11 ++++-----
 .../src/devices/v4l2/v4l2_ll_biases.cpp       | 24 +++++++++++++++----
 4 files changed, 31 insertions(+), 12 deletions(-)

diff --git a/hal_psee_plugins/psee_hw_layer_headers/include/metavision/psee_hw_layer/devices/v4l2/v4l2_ll_biases.h b/hal_psee_plugins/psee_hw_layer_headers/include/metavision/psee_hw_layer/devices/v4l2/v4l2_ll_biases.h
index 4bdc465..a28c2a5 100644
--- a/hal_psee_plugins/psee_hw_layer_headers/include/metavision/psee_hw_layer/devices/v4l2/v4l2_ll_biases.h
+++ b/hal_psee_plugins/psee_hw_layer_headers/include/metavision/psee_hw_layer/devices/v4l2/v4l2_ll_biases.h
@@ -17,17 +17,19 @@
 #include <vector>
 
 #include "metavision/hal/facilities/i_ll_biases.h"
+#include "metavision/hal/facilities/i_hw_identification.h"
 #include "metavision/psee_hw_layer/devices/common/bias_settings.h"
 #include "metavision/psee_hw_layer/boards/v4l2/v4l2_controls.h"
 
 namespace Metavision {
 
 class RegisterMap;
+class I_HW_Identification;
 class V4L2Controls;
 
 class V4L2LLBiases : public I_LL_Biases {
 public:
-    V4L2LLBiases(const DeviceConfig &device_config, std::shared_ptr<V4L2Controls> controls, bool relative = false);
+    V4L2LLBiases(const DeviceConfig &device_config, const std::shared_ptr<I_HW_Identification> &hw_identification, std::shared_ptr<V4L2Controls> controls, bool relative = false);
 
     virtual std::map<std::string, int> get_all_biases() const override;
 
@@ -57,8 +59,10 @@ private:
     virtual bool get_bias_info_impl(const std::string &bias_name, LL_Bias_Info &info) const override;
 
     std::shared_ptr<RegisterMap> register_map_;
+    std::shared_ptr<I_HW_Identification> hw_identification_;
     int fd_;
     bool relative_;
+    int bias_bitwidth;
 };
 
 } // namespace Metavision
diff --git a/hal_psee_plugins/src/boards/v4l2/v4l2_hardware_identification.cpp b/hal_psee_plugins/src/boards/v4l2/v4l2_hardware_identification.cpp
index 81e3501..f7125a4 100644
--- a/hal_psee_plugins/src/boards/v4l2/v4l2_hardware_identification.cpp
+++ b/hal_psee_plugins/src/boards/v4l2/v4l2_hardware_identification.cpp
@@ -91,6 +91,6 @@ std::string V4l2HwIdentification::get_connection_type() const {
 }
 
 DeviceConfigOptionMap V4l2HwIdentification::get_device_config_options_impl() const {
-    return {};
+    return {{"ll_biases_range_check_bypass", DeviceConfigOption(true)}};
 }
 }
diff --git a/hal_psee_plugins/src/devices/v4l2/v4l2_device_builder.cpp b/hal_psee_plugins/src/devices/v4l2/v4l2_device_builder.cpp
index 85ecf92..820651a 100644
--- a/hal_psee_plugins/src/devices/v4l2/v4l2_device_builder.cpp
+++ b/hal_psee_plugins/src/devices/v4l2/v4l2_device_builder.cpp
@@ -54,12 +54,11 @@ bool V4L2DeviceBuilder::build_device(std::shared_ptr<BoardCommand> cmd, DeviceBu
 
     if (controls->has("bias")) {
         MV_HAL_LOG_TRACE() << "Found BIAS controls\n";
-        auto sensor_info = hw_identification->get_sensor_info();
-        bool relative    = false;
-        if (sensor_info.name_ == "IMX636") {
-            relative = true;
-        }
-        device_builder.add_facility(std::make_unique<V4L2LLBiases>(config, controls, relative));
+
+	DeviceConfig v4l2_config = config;
+	v4l2_config.enable_biases_range_check_bypass(true);
+
+        device_builder.add_facility(std::make_unique<V4L2LLBiases>(v4l2_config, hw_identification, controls));
     }
 
     if (controls->has("erc")) {
diff --git a/hal_psee_plugins/src/devices/v4l2/v4l2_ll_biases.cpp b/hal_psee_plugins/src/devices/v4l2/v4l2_ll_biases.cpp
index c698ac6..113d151 100644
--- a/hal_psee_plugins/src/devices/v4l2/v4l2_ll_biases.cpp
+++ b/hal_psee_plugins/src/devices/v4l2/v4l2_ll_biases.cpp
@@ -10,14 +10,24 @@
  **********************************************************************************************************************/
 
 #include <cassert>
+#include <cmath>
 #include "metavision/hal/utils/detail/hal_log_impl.h"
 #include "metavision/psee_hw_layer/devices/v4l2/v4l2_ll_biases.h"
 #include "metavision/psee_hw_layer/boards/v4l2/v4l2_controls.h"
 
 namespace Metavision {
 
-V4L2LLBiases::V4L2LLBiases(const DeviceConfig &device_config, std::shared_ptr<V4L2Controls> controls, bool relative) :
-    I_LL_Biases(device_config), controls_(controls), relative_(relative) {
+V4L2LLBiases::V4L2LLBiases(const DeviceConfig &device_config, const std::shared_ptr<I_HW_Identification> &hw_identification, std::shared_ptr<V4L2Controls> controls, bool relative) :
+    I_LL_Biases(device_config), hw_identification_(hw_identification), controls_(controls) { 
+    auto sensor_info = hw_identification->get_sensor_info();
+    if (sensor_info.name_ == "IMX636") {
+        relative_ = true;
+        bias_bitwidth = 8;
+    } else {
+        relative_ = false;
+        bias_bitwidth = 7;
+    }
+
     // reset all biases to default values
     controls_->foreach ([&](V4L2Controls::V4L2Control &ctrl) {
         auto name = std::string(ctrl.query_.name);
@@ -62,8 +72,14 @@ int V4L2LLBiases::get_impl(const std::string &bias_name) const {
 
 bool V4L2LLBiases::get_bias_info_impl(const std::string &bias_name, LL_Bias_Info &bias_info) const {
     auto ctrl = controls_->get(bias_name);
-    bias_info = LL_Bias_Info(0, 127, ctrl.query_.minimum, ctrl.query_.maximum, std::string("todo::description"), true,
-                             std::string("todo::category"));
+    int offset = 0;
+
+    if (relative_) {
+	offset = ctrl.query_.default_value;
+    }
+
+    bias_info = LL_Bias_Info(0 - offset, (pow(2, bias_bitwidth) - 1) - offset, ctrl.query_.minimum - offset, ctrl.query_.maximum - offset,
+                             std::string("todo::description"), true, std::string("todo::category"));
 
     return true;
 }
